src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -0,0 +1,783 @@
+            /// <returns>A status indicating whether the conversion succeeded and why</returns>
+            public SLD_STATUS AssembleFloatingPointValue(
+                ulong initialMantissa,
+                int initialExponent,
+                bool hasZeroTail,
+                out ulong result)
+            {
+                // number of bits by which we must adjust the mantissa to shift it into the  
+                // correct position, and compute the resulting base two exponent for the  
+                // normalized mantissa:  
+                uint initialMantissaBits = CountSignificantBits(initialMantissa);
+                int normalMantissaShift = this.NormalMantissaBits - (int)initialMantissaBits;
+                int normalExponent = initialExponent - normalMantissaShift;
+
+                ulong mantissa = initialMantissa;
+                int exponent = normalExponent;
+
+                if (normalExponent > this.MaxBinaryExponent)
+                {
+                    // The exponent is too large to be represented by the floating point  
+                    // type; report the overflow condition:
+                    result = this.Infinity;
+                    return SLD_STATUS.SLD_OVERFLOW;
+                }
+                else if (normalExponent < this.MinBinaryExponent)
+                {
+                    // The exponent is too small to be represented by the floating point  
+                    // type as a normal value, but it may be representable as a denormal  
+                    // value.  Compute the number of bits by which we need to shift the  
+                    // mantissa in order to form a denormal number.  (The subtraction of  
+                    // an extra 1 is to account for the hidden bit of the mantissa that  
+                    // is not available for use when representing a denormal.)  
+                    int denormalMantissaShift =
+                        normalMantissaShift +
+                        normalExponent +
+                        this.ExponentBias -
+                        1;
+
+                    // Denormal values have an exponent of zero, so the debiased exponent is  
+                    // the negation of the exponent bias:  
+                    exponent = -this.ExponentBias;
+
+                    if (denormalMantissaShift < 0)
+                    {
+                        // Use two steps for right shifts:  for a shift of N bits, we first  
+                        // shift by N-1 bits, then shift the last bit and use its value to  
+                        // round the mantissa.  
+                        mantissa = RightShiftWithRounding(mantissa, -denormalMantissaShift, hasZeroTail);
+
+                        // If the mantissa is now zero, we have underflowed:  
+                        if (mantissa == 0)
+                        {
+                            result = this.Zero;
+                            return SLD_STATUS.SLD_UNDERFLOW;
```

Comment:
So, as a caller, I can (or, rather, I must) then choose whether or not to rationally interpret this as a zero if appropriate for my scenario, correct?  (People consider overflow a lot -- underflow often slips their minds...)